tracpro/polls/tasks.py

```diff
@@ -20,102 +21,59 @@ def fetch_all_runs():
     """
     Fetches flow runs for all orgs
     """
-    logger.info("Starting flow run fetch for all orgs...")
+    r = get_redis_connection()
+
+    # only do this if we aren't already running so we don't get backed up
+    if not r.get(FETCH_ALL_RUNS_LOCK):
+        with r.lock(FETCH_ALL_RUNS_LOCK, timeout=600):
+            logger.info("Starting flow run fetch for all orgs...")
 
-    for org in Org.objects.filter(is_active=True).prefetch_related('polls'):
-        fetch_org_runs(org)
+            for org in Org.objects.filter(is_active=True).prefetch_related('polls'):
+                fetch_org_runs(org)
+    else:
+        logger.warn("Skipping run fetch as it is already running")
 
 
 def fetch_org_runs(org):
     """
-    Fetches flow runs for the given org
+    Fetches new and modified flow runs for the given org and creates/updates poll responses
     """
     from tracpro.orgs_ext.constants import TaskType
-
-    fetch_org_new_runs(org)
-    fetch_org_updated_runs(org)
-
-    task_result = dict(time=datetime_to_ms(timezone.now()))
-    org.set_task_result(TaskType.fetch_runs, task_result)
-
-
-def fetch_org_new_runs(org):
     from tracpro.polls.models import Poll, Response
 
     client = org.get_temba_client()
     r = get_redis_connection()
     last_time_key = LAST_FETCHED_RUN_TIME_KEY % org.pk
-
-    newest_run = None
     last_time = r.get(last_time_key)
+
     if last_time is not None:
         last_time = parse_iso8601(last_time)
     else:
         newest_run = Response.objects.filter(pollrun__poll__org=org).order_by('-created_on').first()
         last_time = newest_run.created_on if newest_run else None
 
-    polls_by_flow_uuids = {p.flow_uuid: p for p in Poll.get_all(org)}
-
-    runs = client.get_runs(flows=polls_by_flow_uuids.keys(), after=last_time)
+    until = timezone.now()
 
-    if runs:
-        # because the Temba API compares after dates with gte, oldest run will be usually be a duplicate
-        oldest_run = runs[len(runs) - 1]
-        if Response.objects.filter(flow_run_id=oldest_run.id).exists():
-            runs = runs[0:-1]
-
-    logger.info("Fetched %d new runs for org #%d (after=%s)"
-                % (len(runs), org.id, format_iso8601(last_time) if last_time else 'Never'))
-
-    if runs:
-        newest_run = runs[0]
+    total_runs = 0
+    for poll in Poll.get_all(org):
+        poll_runs = client.get_runs(flows=[poll.flow_uuid], after=last_time, before=until)
```

Comment:
It's very likely at some point in the future, for performance reasons, we'll have to restrict the API to fetching runs by a single flow, or for all org flows... so future proofing this now.